common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java

```diff
@@ -366,59 +418,112 @@ private void validateEncodeVersion(String encryptedBlob, int encodeVersionLength
+                AZURE_AUTHENTICATOR_APP_PACKAGE_NAME.equalsIgnoreCase(packageName)) {
+            return loadSecretKeyForBrokerEncryption();
+        }
+
+        return loadSecretKeyForADALMSALEncryption();
     }
 
-    @Nullable
-    private byte[] getSecretKeyData(@Nullable final String pkgName) {
-        byte[] secretKeyData;
+    /**
+     * Load SecretKey for Broker for encryption.
+     * This will try getting Keystore-encrypted symmetric key in the following order.
+     * 1. Keystore-encrypted symmetric key.
+     * 2. a newly-generated key.
+     */
+    private SecretKey loadSecretKeyForBrokerEncryption() throws IOException, GeneralSecurityException {
+        setBlobVersion(VERSION_ANDROID_KEY_STORE);
+        SecretKey key = loadSecretKey(KeyType.KEYSTORE_ENCRYPTED_KEY);
 
-        if (AuthenticationSettings.INSTANCE.getBrokerSecretKeys().containsKey(pkgName)) {
-            // The current app runtime is the broker; load its secret keys...
-            secretKeyData = AuthenticationSettings.INSTANCE.getBrokerSecretKeys().get(pkgName);
-        } else { // We are not the broker, proceed as usual.
-            secretKeyData = AuthenticationSettings.INSTANCE.getSecretKeyData();
+        if (key == null) {
+            key = createKey();
         }
 
-        return secretKeyData;
+        return key;
     }
 
-    @Override
-    public synchronized SecretKey loadSecretKeyForEncryption(String defaultBlobVersion) throws IOException,
-            GeneralSecurityException {
-        // Loading key only once for performance. If API is upgraded, it will
-        // restart the device anyway. It will load the correct key for new API.
-        if (mKey != null && mHMACKey != null) {
-            return mKey;
+    /**
+     * Load SecretKey for ADAL/MSAL for encryption.
+     * This will try getting key in the following order.
+     * 1. User-defined key.
+     * 2. Keystore-encrypted symmetric key.
+     */
+    private SecretKey loadSecretKeyForADALMSALEncryption() throws IOException, GeneralSecurityException {
+        if (AuthenticationSettings.INSTANCE.getSecretKeyData() != null) {
+            setBlobVersion(VERSION_USER_DEFINED);
+            return loadSecretKey(KeyType.ADAL_USER_DEFINED_KEY);
+        }
+
+        // Try loading existing keystore-encrypted key. If it doesn't exist, create a new one.
+        setBlobVersion(VERSION_ANDROID_KEY_STORE);
+        SecretKey key = loadSecretKey(KeyType.KEYSTORE_ENCRYPTED_KEY);
+
+        if (key == null) {
+            key = createKey();
+        }
+
+        return key;
+    }
+
+    /**
+     * A function for setting mBlobVersion.
+     * Exposed for test cases.
+     */
+    protected void setBlobVersion(@NonNull String blobVersion) {
+        mBlobVersion = blobVersion;
+    }
+
+    /**
+     * Given the key type, load a secret key.
+     */
+    protected SecretKey loadSecretKey(@NonNull final KeyType keyType) throws IOException, GeneralSecurityException {
```

Comment:
nit: `@NonNull` annotation could be added here.

Reply:
This could be nullable.